### DISCUSSION TITLE
Render two volumes with opacity

### DISCUSSION THREAD

**Discussion by danielfails** [pyvista/pyvista#4384] [Q&A] [answered]
Hi Everyone,

i try to render two overlapping volumes. In my toy-example, the volumes only conist of 0 and 255 values. 0 values should have 0 opacity and 255 values should have 1.0 opacity respectively. See my code snipped below.

I made a smal gif of the rendered volumes. As i rotate the volumes, i expect the yellow volume to hide the red one, as i have set the "yellow opacity to 1.0".

Either I have a fundamental misunderstanding here, it's a bug, or there's another way I can accomplish this.

I am happy for every suggestion :-)

I am working with Pyvista-version: 0.38.6 on a Windows 11 machine.

```python
import numpy as np

import pyvista as pv

volume1 = np.zeros(shape=(100,100,100), dtype=np.float32)
volume1[:,0:50,:] = 255

volume2 = np.zeros(shape=(100,100,100), dtype=np.float32)
volume2[:,50:,:] = 255

p = pv.Plotter()
p.add_volume(volume=volume1, opacity=[0.0, 1.0], cmap="spring")
p.add_volume(volume=volume2, opacity=[0.0, 1.0], cmap="Reds")
p.show()

```
![pyvista](https://user-images.githubusercontent.com/132590594/236249861-b5ff48f9-44c0-42de-a946-9fb0019ad270.gif)

**Comment by banesullivan** [comment]
I'm not sure you want to use volume rendering at all here... The data are constant values so volume rendering isn't going to help here. Volume rendering is typically used to see internal variation in the data. Having two volumes overlapping with wildly different opacities is causing fighting to happen, thus messing up the visualization as you describe.

I'd visualize this like:

```py
import numpy as np
import pyvista as pv

volume = np.zeros(shape=(100,100,100), dtype=np.float32)
volume[:,0:50,:] = 255
# volume[:,50:,:] = 0

p = pv.Plotter()
p.add_mesh(volume, cmap=["Yellow", "Red"])
p.show()
```

![screenshot (3)](https://user-images.githubusercontent.com/22067021/236280874-4b4aa8a3-cd32-413c-b4aa-6163d77e8089.png)


Or if you must volume render:

```py
volume = np.zeros(shape=(100,100,100), dtype=np.float32)
volume[:,0:50,:] = 255
# volume[:,50:,:] = 0

p = pv.Plotter()
p.add_volume(volume, opacity=[0.5, 1.0, 0.5], cmap=["Yellow", "Red"])
p.show()
```
![screenshot (2)](https://user-images.githubusercontent.com/22067021/236280790-bf726621-68c4-4355-a2e5-5adcf88d1857.png)



Hope this helps

**Reply by danielfails** [reply]
Hi, 
Thanks for your reply. The data I want to visualize in my real example is not binary, so using mesh rendering is quite challenging.

I will check if I can merge my volumes into a single one as you suggested and make it look nicely by defining colors myself. Thanks for that.

However, is my expectation correct, that the yellow volume should hide the red one at the end of my gif or do I have a misunderstanding of what it means to render two overlapping volumes?

**Reply by danielfails** [reply]
I added all volumes together into a single one and rendered it. Here is the problem:
In my real world example I have three volumes. Volume A describes the spatial domain (mask with 1s and 0s). Volume B describes a special region within the spatial domain (mask with 1s and 0s). Volume C describes the soluation of a PDE on the given spatial domain. If add together these three volumes, the final rendered volume looks ok, but the problem is, that Values of Volume A and Volume B are present in the colormap legend in the final rendering. That is however completely wrong from the application point of view. The only purpose of Volume B is to visualize a special region within the spatial domain.
Therefore, adding one single volume or mesh is not applicable to me.

If I want to render the three volumes as meshes I run into the problem of setting the 0s of every volume to 0 opacity. In the toy example i want to do something like this:
```python

import numpy as np
import pyvista as pv

volume = np.zeros(shape=(100,100,100), dtype=np.float32)
volume[:,0:50,:] = 255

p = pv.Plotter()
p.add_mesh(volume, cmap=["Yellow", "Red"], opacity=[0.2, 1.0], n_colors=2)
p.show()
```

I guess I really need to render the volumes seperately as given in the toy example and it works but not for all camera positions :-(
This however, renders everything invisible or gives me "RuntimeError: Transfer function cannot have more values than `n_colors`. This has 256 elements" if I don't set "n_colors=2". :-(

**Reply by banesullivan** [reply]
Quickly read this, and my first thought is maybe you just need to toy with the `clim`? Otherwise I will read this again in detail and try to make a similar example to your real use case when I have a moment

**Reply by danielfails** [reply]
Thanks for you effort. I adopted the toy example, to be more like my real world application. Hope this helps.

```python
import numpy as np

import pyvista as pv

volume1 = np.zeros(shape=(100,100,100), dtype=np.float32) # spatial domain, yellow
volume1[10:90,10:90,10:90] = 255

volume2 = np.zeros(shape=(100,100,100), dtype=np.float32) # solution of pde on spatial domain, red
volume2[10:90,60:90,10:90] = 255

volume3 = np.zeros(shape=(100,100,100), dtype=np.float32) # special region within spatial domain, blue
volume3[10:50,60:70,10:50] = 255

p = pv.Plotter()
p.add_volume(volume=volume1, opacity=[0.0, 0.05], cmap="spring")
p.add_volume(volume=volume2, opacity=[0.0, 1.0], cmap="Reds")
p.add_volume(volume=volume3, opacity=[0.0, 1.0], cmap="Blues")
p.show()

```

From this point of view i expect the red volume to cover the blue one as i set the opacity to 1.0 for the red one.

![pyvista_3](https://user-images.githubusercontent.com/132590594/236766660-670c30b8-ba4b-4a80-be6f-07a3600db79f.PNG)

**Reply by danielfails** [reply, marked as answer]
I was actually able to produce what i wanted using the following code for my toy example:

```python
import numpy as np

import pyvista as pv

def np_to_grid(array):
    # Create the spatial reference
    grid = pv.UniformGrid()
    grid.dimensions = np.array(array.shape) + 1
    grid.spacing = (1,1,1)  # These are the cell sizes along each axis
    grid.cell_data["values"] = array.flatten(order="F")  # Flatten the array
    return grid.threshold_percent(percent=0.5).extract_geometry().smooth(n_iter=10)

volume1 = np.zeros(shape=(100,100,100), dtype=np.float32) # spatial domain, yellow
volume1[10:90,10:90,10:90] = 255

volume2 = np.zeros(shape=(100,100,100), dtype=np.float32) # solution of pde on spatial domain, red
volume2[10:90,60:90,10:90] = 255

volume3 = np.zeros(shape=(100,100,100), dtype=np.float32) # special region within spatial domain, blue
volume3[10:50,60:70,10:50] = 255

p = pv.Plotter()
p.add_mesh(mesh=np_to_grid(volume1), opacity=0.05, cmap="spring", clim=(0,255), show_scalar_bar=False)
p.add_mesh(mesh=np_to_grid(volume3), opacity=1.0, cmap="Blues", clim=(0,255), show_scalar_bar=False)
p.add_mesh(mesh=np_to_grid(volume2), opacity=0.99, cmap="Reds", clim=(0,255), show_scalar_bar=True)

p.show()
```

As soon as we've published the pics of my real world example, i will share the pics here.
I still believe, that there is a bug in the multi-volume rendering example i posted in the first place. 

But thanks alot @banesullivan, your hints helped me to try a couple of things out and accomplish this task :-)
